diff --git a/src/SSLSocket.c b/src/SSLSocket.c
index 3b4af66..0d28e22 100644
--- a/src/SSLSocket.c
+++ b/src/SSLSocket.c
@@ -706,6 +706,19 @@ exit:
 	return rc;
 }
 
+int isIpv4or6(char const *host)
+{
+	struct addrinfo *addr = NULL;
+	struct addrinfo hints = {0};
+
+	hints.ai_family=PF_UNSPEC;
+	hints.ai_flags=AI_NUMERICHOST;
+	int match = getaddrinfo (host, NULL, &hints, &addr) == 0;
+	if(addr != NULL)
+		freeaddrinfo(addr);
+
+	return match;
+}
 
 int SSLSocket_setSocketForSSL(networkHandles* net, MQTTClient_SSLOptions* opts,
 	const char* hostname, size_t hostname_len)
@@ -740,15 +753,19 @@ int SSLSocket_setSocketForSSL(networkHandles* net, MQTTClient_SSLOptions* opts,
 			else
 				SSLSocket_error("SSL_set_fd", net->ssl, net->socket, rc, NULL, NULL);
 		}
+
 		hostname_plus_null = malloc(hostname_len + 1u );
 		if (hostname_plus_null)
 		{
-			MQTTStrncpy(hostname_plus_null, hostname, hostname_len + 1u);
-			if ((rc = SSL_set_tlsext_host_name(net->ssl, hostname_plus_null)) != 1) {
-				if (opts->struct_version >= 3)
-					SSLSocket_error("SSL_set_tlsext_host_name", NULL, net->socket, rc, opts->ssl_error_cb, opts->ssl_error_context);
-				else
-					SSLSocket_error("SSL_set_tlsext_host_name", NULL, net->socket, rc, NULL, NULL);
+			MQTTStrncpy(hostname_plus_null, hostname, hostname_len + 1u);
+			if(opts->struct_version > 5 || !isIpv4or6(hostname_plus_null))
+			{
+				if ((rc = SSL_set_tlsext_host_name(net->ssl, hostname_plus_null)) != 1) {
+					if (opts->struct_version >= 3)
+						SSLSocket_error("SSL_set_tlsext_host_name", NULL, net->socket, rc, opts->ssl_error_cb, opts->ssl_error_context);
+					else
+						SSLSocket_error("SSL_set_tlsext_host_name", NULL, net->socket, rc, NULL, NULL);
+				}
 			}
 			free(hostname_plus_null);
 		}
